### PR TITLE
[JENKINS-34246] Make AbstractWorkflowMultiBranchProjectFactory handle deletion or update of existing children

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/branch-api-plugin/pull/45.

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -84,7 +84,7 @@ THE SOFTWARE.
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>branch-api</artifactId>
-            <version>1.9</version>
+            <version>1.10-SNAPSHOT</version> <!-- TODO https://github.com/jenkinsci/branch-api-plugin/pull/45 -->
```
Comment (jglick) [inline pom.xml]: The actual fix is just picking up the dependency.

File: src/main/java/org/jenkinsci/plugins/workflow/multibranch/AbstractWorkflowMultiBranchProjectFactory.java [hunk 2]
```diff
@@ -47,6 +50,12 @@
         return project;
     }
 
+    @Override public final void updateExistingProject(MultiBranchProject<?, ?> project, Map<String, Object> attributes, TaskListener listener) throws IOException, InterruptedException {
+        if (project instanceof WorkflowMultiBranchProject) {
+            customize((WorkflowMultiBranchProject) project);
+        } // otherwise got recognized by something else before, oh well
+    }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/multibranch/AbstractWorkflowMultiBranchProjectFactory.java]: For use by non-`WorkflowMultiBranchProjectFactory` subclasses.

File: src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactoryTest.java [hunk 3]
```diff
@@ -115,6 +115,12 @@
         b1 = p.getLastBuild();
         assertEquals(1, b1.getNumber());
         r.assertLogContains("ran two", b1);
+        // JENKINS-34246: also delete Jenkinsfile
+        sampleRepo2.git("rm", WorkflowBranchProjectFactory.SCRIPT);
+        sampleRepo2.git("commit", "--message=noflow");
+        top.scheduleBuild2(0).getFuture().get();
+        top.getComputation().writeWholeLogTo(System.out);
+        assertEquals(1, top.getItems().size());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowMultiBranchProjectFactoryTest.java]: The integration test.

**Comment by amuniz** [comment]
:bee:

**Pull request merged**